### PR TITLE
terraform: return `initialization required` error when provider schemas not found

### PR DESCRIPTION
**Pull request by mildwonkey**
A side effect of the various changes to the provider installer included losing the `initialization required` error message which would occur if a user removed or modified the `.terraform` directory.

Previously, plugin factories were created after the configuration was loaded, in `terraform.NewContext`.  Terraform would compare the required providers (from config and state) to the available providers and return the aforementioned error if a provider was missing.

Provider factories are now loaded at the beginning of any terraform command, before terraform even loads the configuration, and therefore before terraform has a list of required providers. 

This PR replaces the current error when a providers' schema cannot be found in the provider factories with the `init` error, and adds a command test (to `plan` tests, for no real reason other than it's what I thought of first).

### CODE REVIEW

**Comment by mildwonkey** [comment]
Here's the output:

Screen Shot 2020-04-20 at 3 12 27 PM

File: terraform/context.go [hunk 1]
```diff
@@ -177,7 +177,11 @@ func NewContext(opts *ContextOpts) (*Context, tfdiags.Diagnostics) {
 	log.Printf("[TRACE] terraform.NewContext: loading provider schemas")
 	schemas, err := LoadSchemas(opts.Config, opts.State, components)
 	if err != nil {
-		diags = diags.Append(err)
+		diags = diags.Append(tfdiags.Sourceless(
+			tfdiags.Error,
+			"Could not satisfy plugin requirements",
```
Comment (alisdair) [inline terraform/context.go:182]: I think this is fine, but "Failed to load plugins" or "Could not load plugins" might be reasonable alternatives.

Comment (mildwonkey) [reply terraform/context.go:182]: I like it! You've given me a lot of great suggestions for logging/error messages etc lately! 
_makes a note to ask @alisdair to write all my logging from now on_

File: terraform/resource_provider.go [hunk 2]
```diff
@@ -231,4 +231,6 @@ don't satisfy the version constraints, or are otherwise incompatible.
 Terraform automatically discovers provider requirements from your
 configuration, including providers used in child modules. To see the
 requirements and constraints from each module, run "terraform providers".
+
```
Comment (apparentlymart) [inline terraform/resource_provider.go:234]: Minor nitpick: the error message above is hinting at some functionality of `terraform providers` that we recently removed. It now just prints out a flat set of constraints, not _constraints from each module_.

It's written vaguely enough that we could consider it still technically correct, but maybe dropping " from each module" from that sentence would be more honest? :man_shrugging:

Comment (mildwonkey) [reply terraform/resource_provider.go:234]: I'm glad you caught that, thank you, I've made that update.

**Pull request merged**